File: app/src/main/java/com/nruge/iceinfo/ui/components/TravelSummaryCard.kt
package com.nruge.iceinfo.ui.components

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.KeyboardArrowUp
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.drawscope.withTransform
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.drawText
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.nruge.iceinfo.R
import com.nruge.iceinfo.model.TrainStatus
import com.nruge.iceinfo.util.formatRemainingTimeUntil

@OptIn(ExperimentalMaterial3ExpressiveApi::class)
@Composable
fun TravelSummaryCard(status: TrainStatus) {
    val targetStop = status.stops.find { it.evaNr == status.targetStopEva }
    val displayDestination = targetStop?.name ?: status.destination
    val displayEta = targetStop
        ?.let { it.actualArrival.takeIf { t -> t.isNotEmpty() } ?: it.scheduledArrival }
        ?: status.destinationEta
    val displayDelay = targetStop?.delayMinutes ?: status.destinationDelay

    val stopsToTarget = if (targetStop != null) {
        status.stops.filter { !it.passed && it.distanceFromStart <= targetStop.distanceFromStart }
    } else {
        status.stops.filter { !it.passed }
    }

    val totalStopsInJourney = status.stops.size
    val passedStops = status.stops.count { it.passed }

    val targetDistance = targetStop?.distanceFromStart ?: status.stops.lastOrNull()?.distanceFromStart ?: 0
    val currentPosition = status.actualPosition
    val remainingDistanceToTarget = (targetDistance - currentPosition).coerceAtLeast(0)

    val totalDistanceForProgress = targetDistance.toFloat()
    val progressPercent = if (totalDistanceForProgress > 0) {
        (currentPosition.toFloat() / totalDistanceForProgress).coerceIn(0f, 1f)
    } else 0f

    val primaryColor = MaterialTheme.colorScheme.primary

    val endStop = targetStop ?: status.stops.lastOrNull { !it.passed && !it.isCancelled }
    val intermediateDotStops = stopsToTarget.filter {
        it.distanceFromStart > currentPosition && !it.isCancelled && it != endStop
    }
    val labelStops = buildList {
        addAll(intermediateDotStops)
        if (endStop != null) add(endStop)
    }

    val textMeasurer = rememberTextMeasurer()
    var showStopLabels by remember { mutableStateOf(true) }

    AppCard(modifier = Modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            // Destination + label toggle
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = " ➜ $displayDestination",
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.Bold,
                    color = primaryColor,
                    modifier = Modifier.weight(1f)
                )
                if (labelStops.isNotEmpty()) {
                    IconButton(
                        onClick = { showStopLabels = !showStopLabels },
                        modifier = Modifier.size(28.dp)
                    ) {
                        Icon(
                            imageVector = if (showStopLabels) Icons.Default.KeyboardArrowUp
                                          else Icons.Default.KeyboardArrowDown,
                            contentDescription = if (showStopLabels) "Halte ausblenden"
                                                 else "Halte einblenden",
                            tint = primaryColor.copy(alpha = 0.6f),
                            modifier = Modifier.size(18.dp)
                        )
                    }
                }
            }

            Spacer(modifier = Modifier.height(7.dp))

            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    text = stringResource(R.string.travel_remaining_km, remainingDistanceToTarget / 1000),
                    style = MaterialTheme.typography.bodyMedium,
                    color = primaryColor
                )
                Text(
                    text = if (targetStop != null)
                        stringResource(R.string.travel_remaining_stops, labelStops.size)
                    else
                        stringResource(R.string.travel_stops_progress, passedStops, totalStopsInJourney),
                    style = MaterialTheme.typography.bodySmall,
                    color = primaryColor.copy(alpha = 0.7f)
                )
            }

            // Progress bar with stop dots
            Box(modifier = Modifier.fillMaxWidth()) {
                if (status.speed > 0) {
                    LinearWavyProgressIndicator(
                        progress = { progressPercent },
                        modifier = Modifier.fillMaxWidth(),
                        color = primaryColor,
                        trackColor = primaryColor.copy(alpha = 0.2f),
                        waveSpeed = (status.speed / 15f).coerceAtMost(20f).dp
                    )
                } else {
                    LinearProgressIndicator(
                        progress = { progressPercent },
                        modifier = Modifier.fillMaxWidth(),
                        color = primaryColor,
                        trackColor = primaryColor.copy(alpha = 0.2f)
                    )
                }
                if (totalDistanceForProgress > 0 && labelStops.isNotEmpty()) {
                    Canvas(modifier = Modifier.matchParentSize()) {
                        intermediateDotStops.forEach { stop ->
                            val fraction = (stop.distanceFromStart.toFloat() / totalDistanceForProgress)
                                .coerceIn(0f, 1f)
                            drawCircle(
                                color = if (stop.isNext) primaryColor else primaryColor.copy(alpha = 0.5f),
                                radius = if (stop.isNext) 5.dp.toPx() else 3.5.dp.toPx(),
                                center = Offset(fraction * size.width, size.height / 2f)
                            )
                        }
                        if (endStop != null) {
                            drawCircle(
                                color = primaryColor,
                                radius = 5.dp.toPx(),
                                center = Offset(size.width, size.height / 2f)
                            )
                        }
                    }
                }
            }

            // Station name labels — vertically below each dot, centered on the dot's x position
            if (showStopLabels && totalDistanceForProgress > 0 && labelStops.isNotEmpty()) {
                val labelStyle = TextStyle(fontSize = 9.sp, color = primaryColor.copy(alpha = 0.65f))
                val labelStyleBold = TextStyle(fontSize = 9.sp, color = primaryColor)
                Canvas(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(70.dp)
                ) {
                    labelStops.forEach { stop ->
                        val isEndpoint = stop == endStop
                        val style = if (stop.isNext || isEndpoint) labelStyleBold else labelStyle
                        val measured = textMeasurer.measure(stop.name, style)

                        val dotX = if (isEndpoint) size.width
                                   else (stop.distanceFromStart.toFloat() / totalDistanceForProgress)
                                       .coerceIn(0f, 1f) * size.width

                        val labelYOffset = 0.dp.toPx()
                        withTransform({
                            translate(left = dotX, top = labelYOffset)
                            rotate(degrees = -45f, pivot = Offset.Zero)
                        }) {
                            // Right edge of text anchored at dot → text hangs down-left
                            drawText(measured, topLeft = Offset(-measured.size.width.toFloat() - 8.dp.toPx(), 0f))
                        }
                    }
                }
            }

            Spacer(modifier = Modifier.height(7.dp))
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                    Text(
                        text = stringResource(R.string.travel_arrival, displayEta),
                        style = MaterialTheme.typography.bodyMedium,
                        fontWeight = FontWeight.SemiBold,
                        color = primaryColor
                    )
                    Text(
                        text = stringResource(
                            R.string.travel_remaining_time,
                            formatRemainingTimeUntil(
                                targetStop?.scheduledArrival ?: status.destinationEta,
                                displayDelay
                            )
                        ),
                        style = MaterialTheme.typography.bodySmall,
                        color = primaryColor.copy(alpha = 0.7f)
                    )
                }
                if (displayDelay > 0) {
                    DelayBadge(delayMinutes = displayDelay)
                } else {
                    Surface(
                        color = primaryColor.copy(alpha = 0.1f),
                        shape = MaterialTheme.shapes.small
                    ) {
                        Text(
                            text = stringResource(R.string.travel_on_time),
                            modifier = Modifier.padding(horizontal = 8.dp, vertical = 4.dp),
                            color = primaryColor,
                            style = MaterialTheme.typography.labelMedium,
                            fontWeight = FontWeight.Bold
                        )
                    }
                }
            }
        }
    }
}

File: ios/ICEInfo/Views/Components/TravelSummaryCard.swift
import SwiftUI

struct TravelSummaryCard: View {
    let status: TrainStatus

    var body: some View {
        let targetStop = status.stops.first { $0.evaNr == status.targetStopEva && !$0.passed }
        let displayDestination = targetStop?.name ?? status.destination
        let displayEta = targetStop?.actualArrival.nilIfEmpty ?? targetStop?.scheduledArrival ?? status.destinationEta
        let displayDelay = targetStop?.delayMinutes ?? status.destinationDelay

        let stopsToTarget = if let t = targetStop {
            status.stops.filter { !$0.passed && $0.distanceFromStart <= t.distanceFromStart }
        } else {
            status.stops.filter { !$0.passed }
        }
        let totalStops = status.stops.count
        let passedStops = status.stops.filter(\.passed).count
        let targetDist = targetStop?.distanceFromStart ?? status.stops.last?.distanceFromStart ?? 0
        let remainingDist = max(0, targetDist - status.actualPosition)
        let progress = targetDist > 0 ? min(1, max(0, Float(status.actualPosition) / Float(targetDist))) : 0

        AppCard {
            VStack(alignment: .leading, spacing: 12) {
                Text("\u{2192} \(displayDestination)")
                    .font(.subheadline)
                    .fontWeight(.bold)
                    .foregroundStyle(Color(.systemBackground))

                Spacer().frame(height: 7)

                HStack {
                    Text("\(remainingDist / 1000) km verbleibend")
                        .font(.body)
                    Spacer()
                    if let t = targetStop {
                        Text("\(stopsToTarget.count) Halte")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    } else {
                        Text("\(passedStops)/\(totalStops)")
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                }

                if status.speed > 0 {
                    WavyProgressView(progress: progress, speed: status.speed)
                        .frame(height: 8)
                } else {
                    ProgressView(value: Double(progress))
                        .tint(Color(.systemBackground))
                }

                Spacer().frame(height: 7)

                HStack {
                    VStack(alignment: .leading, spacing: 2) {
                        Text("Ankunft \(displayEta)")
                            .font(.body.weight(.semibold))
                        Text(formatRemainingTimeUntil(scheduledArrival: targetStop?.scheduledArrival ?? status.destinationEta, delayMinutes: displayDelay))
                            .font(.caption)
                            .foregroundStyle(.secondary)
                    }
                    Spacer()
                    if displayDelay > 0 {
                        DelayBadge(delayMinutes: displayDelay)
                    } else {
                        Text("P\u{00FC}nktlich")
                            .font(.caption.bold())
                            .padding(.horizontal, 8)
                            .padding(.vertical, 4)
                            .background(Color(.systemBackground).opacity(0.1))
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                }
            }
            .padding(16)
            .background(Color(red: 0.925, green: 0, blue: 0.086))
            .foregroundStyle(.white)
            .clipShape(RoundedRectangle(cornerRadius: 16))
        }
    }
}

struct WavyProgressView: View {
    let progress: Float
    let speed: Int

    @State private var phase: CGFloat = 0

    var body: some View {
        TimelineView(.animation) { timeline in
            GeometryReader { geo in
                let w = geo.size.width
                let h = geo.size.height
                let waveSpeed = min(CGFloat(speed) / 15, 20)
                let p = CGFloat(progress)

                Canvas { context, size in
                    let waveHeight: CGFloat = 3
                    let path = Path { path in
                        path.move(to: CGPoint(x: 0, y: h))
                        for x in stride(from: 0, through: w, by: 1) {
                            let y = h - waveHeight * sin((x / w + phase) * .pi * 4)
                            if x <= w * p {
                                path.addLine(to: CGPoint(x: x, y: y))
                            } else {
                                path.addLine(to: CGPoint(x: x, y: h))
                            }
                        }
                        path.addLine(to: CGPoint(x: w, y: h))
                        path.closeSubpath()
                    }
                    context.fill(path, with: .color(.white.opacity(0.3)))
                }
                .onChange(of: timeline.date.timeIntervalSince1970) { _, _ in
                    phase += 0.02 * waveSpeed
                }
            }
        }
    }
}

extension String {
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}
